torch/utils/data/dataloader.py

```diff
@@ -121,13 +144,91 @@ def __setattr__(self, attr, val):
         super(DataLoader, self).__setattr__(attr, val)
 
     def __iter__(self):
-        return _DataLoaderIter(self)
+        if self.num_workers == 0:
+            return _SingleProcessDataLoaderIter(self)
+        else:
+            return _MultiProcessingDataLoaderIter(self)
 
     def __len__(self):
-        return len(self.batch_sampler)
+        if self.mode == DataLoaderMode.Iterable:
+            raise NotImplementedError
+        return len(self.index_sampler)
+
+
+class _BaseDataLoaderIter(object):
+    def __init__(self, loader):
+        self.dataset = loader.dataset
+        self.mode = loader.mode
+        self.index_sampler = loader.index_sampler
+        self.num_workers = loader.num_workers
+        self.pin_memory = loader.pin_memory and torch.cuda.is_available()
+        self.timeout = loader.timeout
+        self.convert_fn = loader.convert_fn
+        self.collate_fn = loader.collate_fn
+
+        if self.mode == DataLoaderMode.Iterable:
+            self.index_sampler_iter = None
+        else:
+            self.index_sampler_iter = iter(self.index_sampler)
+
+        if self.mode == DataLoaderMode.Iterable and self.num_workers == 0:
+            self.dataset_iter = iter(self.dataset)
+        else:
+            self.dataset_iter = None
+
+        self.base_seed = torch.empty((), dtype=torch.int64).random_().item()
+
+    def __iter__(self):
+        return self
+
+    def _next_index(self):
+        if self.mode == DataLoaderMode.Iterable:
+            return None
+        else:
+            return next(self.index_sampler_iter)  # may raise StopIteration
+
+    def __next__(self):
+        raise NotImplementedError
+
+    def __len__(self):
+        if self.mode == DataLoaderMode.Iterable:
+            raise NotImplementedError
+        else:
+            return len(self.index_sampler)
+
+    def __getstate__(self):
+        # TODO: add limited pickling support for sharing an iterator
+        # across multiple threads for HOGWILD.
+        # Probably the best way to do this is by moving the sample pushing
+        # to a separate thread and then just sharing the data queue
+        # but signalling the end is tricky without a non-blocking API
+        raise NotImplementedError("_DataLoaderIter cannot be pickled")
+
+
+class _SingleProcessDataLoaderIter(_BaseDataLoaderIter):
+    def __init__(self, loader):
+        super(_SingleProcessDataLoaderIter, self).__init__(loader)
+        assert self.timeout == 0
+        assert self.num_workers == 0
+
+    def __next__(self):
+        if self.mode == DataLoaderMode.Iterable:
+            data = self.convert_fn(next(self.dataset_iter))  # may raise StopIteration
+        else:
+            index = self._next_index()  # may raise StopIteration
+            if self.mode == DataLoaderMode.Map:
+                data = self.convert_fn(self.dataset[index])
+            else:
+                # mode == DataLoaderMode.MapWithBatchedRead:
```

Comment:
Because previous there is no easy way to to unbatch loading from a map-like dataset. One has to use `batch_size=1` and provide a custom collate_fn which is just `lambda x: x[0]`. That feels counterintuitive to me. With all other changes in this PR, this was a simple addition, so I made it.

Reply:
Another reason is to provide better support for the case where the sampler provides a list of indices, and the dataset does bulk loading.